### DISCUSSION TITLE
Installing WinGet and PowerShell Module Programmatically on Windows 10

### DISCUSSION THREAD

**Discussion by norsemangrey** [microsoft/winget-cli#4065] [PowerShell]
Has anyone been able to successfully install WinGet as well as the PowerShell module (Microsoft.WinGet.Client) on a Windows 10 machine programmatically (using command line / PS script)?

I have been trying now for a couple of weeks, but it always results in the PS module giving me this error when I try to run commands like `Get-WinGetPackage` or `Find-WinGetPackage`:

```
An error occurred while connecting to the catalog.
```

Using the `winget` command always works so it does not appear to be anything wrong with the WinGet installation itself (lately I have been using [this](https://github.com/asheroto/winget-install) installer script)

I have tried two different approaches; installing PowerShell 7 first (in PS5 using this MSI), then WinGet and finally the Microsoft.WinGet.Client in PS7 ...  and another where I install WinGet first then PS7 using WinGet in PS5, and finally Microsoft.WinGet.Client in PS7. None of the approaches seems to work.

Has anyone tried the same and experienced similar issues?

**Comment by Andrew-J-Larson** [comment]
The PS module only works if WinGet is installed, and installing the PS module itself doesn't install WinGet with it.

But yes, I was able to create a script to install WinGet (but it doesn't install the PS module `Microsoft.WinGet.Client`). The script should even work on machines booted for the first time right after OOBE, and only hooks into websites trusted/used by Microsoft:
- https://github.com/Andrew-J-Larson/OS-Scripts/blob/main/Windows/Wrapper-Functions/Install-WinGet-Function.ps1

**Reply by Andrew-J-Larson** [reply]
Depending on the version of Windows, or the Desktop App Installer (WinGet), it can try some remediating steps first for certain quirks, before going out and actually grabbing and installing files needed for the WinGet install, and then it will install WinGet itself.

**Reply by Andrew-J-Larson** [reply]
It also tries to be efficient by only downloading the files it needs to complete the install (or forced upgrade).

Additionally, for any required packages that can possibly get new versions, this script dynamically grabs latest versions, so no specific versions of packages are hard coded (e.g. should work "forever" basically, unless underlying URL/JSON/XML formats are changed, then it'll need to be fixed at that point).

**Reply by Andrew-J-Larson** [reply]
For anyone that was having issues with the script, I updated it, and it should be working now.

**Comment by Trenly** [comment]
This should work, since it is what is used in the [SandboxTest script](https://github.com/microsoft/winget-pkgs/commit/8cefcfaf51c39a5895955fc78efeb8f67eb90102) over at winget-pkgs.
```
try {
  Install-PackageProvider -Name NuGet -Force | Out-Null
  Install-Module -Name Microsoft.WinGet.Client -Force -Repository PSGallery | Out-Null
} catch {
  throw "Microsoft.Winget.Client was not installed successfully"
} finally {
  # Check to be sure it acutally installed
  if (-not(Get-Module -ListAvailable -Name Microsoft.Winget.Client)) {
    throw "Microsoft.Winget.Client was not found. Check that the Windows Package Manager PowerShell module was installed correctly."
  }
}
Repair-WinGetPackageManager
```

If you don't need the Try-Catch, it can be simplified down to 3 lines -
```
Install-PackageProvider -Name NuGet -Force | Out-Null
Install-Module -Name Microsoft.WinGet.Client -Force -Repository PSGallery | Out-Null
Repair-WinGetPackageManager
```

 As @Andrew-J-Larson said, you need to install the WinGet client in order to use commands like `Get-WinGetPackage` or `Find-WinGetPackage`. However, you *can* use the PowerShell Module to install the WinGet client with the `Repair-WinGetPackageManager` cmdlet, which will automatically handle any of the dependencies needed.

**Reply by Andrew-J-Larson** [reply]
One thing to note here though is that the `Microsoft.Winget.Client` module requires that it be used in Powershell (not the built-in 5.1 version) to be completely compatible with the module.

If the Powershell 5.1 version must be used, the script I created in https://github.com/microsoft/winget-cli/discussions/4065#discussioncomment-8352604 should help.

**Reply by Trenly** [reply]
The `Repair-WinGetPackageManager` cmdlet works in PowerShell 5.1, though, as it was specifically intended for the purpose of fixing any installation issues with WinGet. So, using the PowerShell module to install winget does work

**Reply by TOXXXX** [reply]
Thanks a lot @Trenly , that script solve my problem! I searched how to install winget and all I got is that it comes with "App Manager" from the Microsoft Store. Although I have downloaded and upgraded it, I was still not able to use winget. And this github version of winget does not give any instruction on how to deploy it.

**Reply by lanwin** [reply]
For me this fails with "Install-Module : A parameter cannot be found that matches parameter name 'Name'." and Windows build in PS. Any idea?

**Comment by denelon** [comment]
The latest method for doing this is by running `Repair-WinGetPackageManager -Force -Latest` in the Microsoft.WinGet.Client PowerShell module. This will work even in Windows PowerShell, but many of the other cmdlets will fail unless they are run under PowerShell 7.

**Comment by kkaminsk** [comment]
> Has anyone been able to successfully install WinGet as well as the PowerShell module (Microsoft.WinGet.Client) on a Windows 10 machine programmatically (using command line / PS script)?
> 
> Has anyone tried the same and experienced similar issues?

I haven't looked at the PowerShell module closely but I am now for other reasons. I go have a revised script over here for Winget installations on devices that don't have it.

https://github.com/kkaminsk/InstallWinget-V2
